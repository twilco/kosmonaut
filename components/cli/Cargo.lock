# This file is automatically @generated by Cargo.
# It is not intended for manual editing.
version = 3

[[package]]
name = "ansi_term"
version = "0.11.0"
source = "registry+https://github.com/rust-lang/crates.io-index"
checksum = "ee49baf6cb617b853aa8d93bf420db2383fab46d314482ca2803b40d5fde979b"
dependencies = [
 "winapi",
]

[[package]]
name = "atty"
version = "0.2.14"
source = "registry+https://github.com/rust-lang/crates.io-index"
checksum = "d9b39be18770d11421cdb1b9947a45dd3f37e93092cbf377614828a319d5fee8"
dependencies = [
 "hermit-abi",
 "libc",
 "winapi",
]

[[package]]
name = "bitflags"
version = "1.2.1"
source = "registry+https://github.com/rust-lang/crates.io-index"
checksum = "cf1de2fe8c75bc145a2f577add951f8134889b4795d47466a54a5c846d691693"

[[package]]
name = "clap"
version = "2.33.3"
source = "registry+https://github.com/rust-lang/crates.io-index"
checksum = "37e58ac78573c40708d45522f0d80fa2f01cc4f9b4e2bf749807255454312002"
dependencies = [
 "ansi_term",
 "atty",
 "bitflags",
 "strsim",
 "textwrap",
 "unicode-width",
 "vec_map",
]

[[package]]
name = "cli"
version = "0.1.0"
dependencies = [
 "clap",
]

[[package]]
name = "hermit-abi"
version = "0.1.18"
source = "registry+https://github.com/rust-lang/crates.io-index"
checksum = "322f4de77956e22ed0e5032c359a0f1273f1f7f0d79bfa3b8ffbc730d7fbcc5c"
dependencies = [
 "libc",
]

[[package]]
name = "libc"
version = "0.2.88"
source = "registry+https://github.com/rust-lang/crates.io-index"
checksum = "03b07a082330a35e43f63177cc01689da34fbffa0105e1246cf0311472cac73a"

[[package]]
name = "strsim"
version = "0.8.0"
source = "registry+https://github.com/rust-lang/crates.io-index"
checksum = "8ea5119cdb4c55b55d432abb513a0429384878c15dde60cc77b1c99de1a95a6a"

[[package]]
name = "textwrap"
version = "0.11.0"
source = "registry+https://github.com/rust-lang/crates.io-index"
checksum = "d326610f408c7a4eb6f51c37c330e496b08506c9457c9d34287ecc38809fb060"
dependencies = [
 "unicode-width",
]

[[package]]
name = "unicode-width"
version = "0.1.8"
source = "registry+https://github.com/rust-lang/crates.io-index"
checksum = "9337591893a19b88d8d87f2cec1e73fad5cdfd10e5a6f349f498ad6ea2ffb1e3"

[[package]]
name = "vec_map"
version = "0.8.2"
source = "registry+https://github.com/rust-lang/crates.io-index"
checksum = "f1bddf1187be692e79c5ffeab891132dfb0f236ed36a43c7ed39f1165ee20191"

[[package]]
name = "winapi"
version = "0.3.9"
source = "registry+https://github.com/rust-lang/crates.io-index"
checksum = "5c839a674fcd7a98952e593242ea400abe93992746761e38641405d28b00f419"
dependencies = [
 "winapi-i686-pc-windows-gnu",
 "winapi-x86_64-pc-windows-gnu",
]

[[package]]
name = "winapi-i686-pc-windows-gnu"
version = "0.4.0"
source = "registry+https://github.com/rust-lang/crates.io-index"
checksum = "ac3b87c63620426dd9b991e5ce0329eff545bccbbb34f3be09ff6fb6ab51b7b6"

[[package]]
name = "winapi-x86_64-pc-windows-gnu"
version = "0.4.0"
source = "registry+https://github.com/rust-lang/crates.io-index"
checksum = "712e227841d057c1ee1cd2fb22fa7e5a5461ae8e48fa2ca79ec42cfc1931183f"
