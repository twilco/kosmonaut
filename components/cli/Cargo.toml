[package]
name = "cli"
version = "0.1.0"
description = "Kosmonaut's CLI"
authors = ["Tyler Wilcock <twilco.o@protonmail.com>"]
edition = "2018"
publish = false

[dependencies]
clap = "2.33.0"
