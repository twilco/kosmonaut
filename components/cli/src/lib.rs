use crate::matches::try_get_bool;
use clap::{App, Arg, ArgMatches, SubCommand};

const DUMP_LAYOUT_CMD_NAME: &str = "dump-layout";
const SIMILARITY_CMD_NAME: &str = "similarity";
const RENDER_INPUT_ARG_NAME: &str = "FILES OR URLS";
const DUMP_LAYOUT_INPUT_ARG_NAME: &str = "FILES";
const SIMILARITY_INPUT_ARG_NAME: &str = "FILES";

pub mod commands;
pub mod matches;

pub fn setup_and_get_cli_args<'a>() -> ArgMatches<'a> {
    let headed_or_headless_applicable =
        "Applicable in both headed and headless (e.g. the dump-layout and similarity commands) contexts.";
    let scale_factor_help = format!(
        "Device/window scale factor.  {}",
        headed_or_headless_applicable
    );
    let scale_factor_arg = Arg::with_name("scale-factor")
        .short("s")
        .long("scale-factor")
        .value_name("NUMBER")
        .help(&scale_factor_help)
        .takes_value(true)
        .validator(is_num_validator);
    let width_help = format!("Window width.  {}", headed_or_headless_applicable);
    let width_arg = Arg::with_name("width")
        .short("w")
        .long("width")
        .value_name("NUMBER")
        .help(&width_help)
        .takes_value(true)
        .validator(is_num_validator);
    let height_help = format!("Window height.  {}", headed_or_headless_applicable);
    let height_arg = Arg::with_name("height")
        .short("h")
        .long("height")
        .value_name("NUMBER")
        .help(&height_help)
        .takes_value(true)
        .validator(is_num_validator);

    App::new("Kosmonaut")
        .version("0.1")
        .author("Tyler Wilcock (twilco)")
        .about("A web browser for the space age ðŸš€")
        .arg(width_arg.clone())
        .arg(height_arg.clone())
        .arg(scale_factor_arg.clone())
        .arg(
            Arg::with_name(RENDER_INPUT_ARG_NAME)
                .help("File(s) or URL(s) for Kosmonaut to render.")
                .index(1)
                .multiple(true)
        )
        .subcommand(
            SubCommand::with_name(DUMP_LAYOUT_CMD_NAME)
                .about("Dumps layout-tree as text to stdout after first global layout, exiting afterwards.")
                .arg(
                    Arg::with_name("verbose")
                        .short("v")
                        .long("verbose")
                        .value_name("BOOLEAN")
                        .help("Set to true to make layout dumps more verbose (e.g. include margin, border, padding values).")
                        .takes_value(true)
                        .validator(is_bool_validator)
                )
                .arg(
                    Arg::with_name(DUMP_LAYOUT_INPUT_ARG_NAME)
                        .help("File(s) for Kosmonaut dump the layout of.  Note only the first HTML file found is rendered, and beyond that only CSS files will be used.")
                        .index(1)
                        .required(true)
                        .min_values(1)
                )
                .arg(scale_factor_arg.clone().required(true))
                .arg(width_arg.clone().required(true))
                .arg(height_arg.clone().required(true))
        )
        .subcommand(
            SubCommand::with_name(SIMILARITY_CMD_NAME)
                .long_about("
Performs a pixel-by-pixel comparison of the renderings of two input HTML files, returning their \
similarity as a percentage.  If you pass any more or any less than two HTML files, this command will \
error.
                ".trim())
                .arg(
                    Arg::with_name("similarity-percent-only")
                        .long("similarity-percent-only")
                        .help("Set to true to make the command only output the similarity percent between the two renderings (e.g. \"99.32%\".")
                )
                .arg(
                    Arg::with_name(DUMP_LAYOUT_INPUT_ARG_NAME)
                        .help("Two HTML files for Kosmonaut to render and compare.")
                        .index(1)
                        .required(true)
                        .min_values(2)
                        .max_values(2)
                )
                .arg(scale_factor_arg)
                .arg(width_arg)
                .arg(height_arg)
        )
        .get_matches()
}

fn is_num_validator(string: String) -> Result<(), String> {
    match string.parse::<f32>() {
        Ok(_) => Ok(()),
        Err(_) => Err(format!("given arg '{}' is not a number", string)),
    }
}

fn is_bool_validator(string: String) -> Result<(), String> {
    match string.parse::<bool>() {
        Ok(_) => Ok(()),
        Err(_) => match &string[..] {
            "0" | "1" => Ok(()),
            _ => Err(format!("given arg '{}' is not a bool value", string)),
        },
    }
}

#[derive(Copy, Clone, Debug, PartialEq, Eq)]
pub enum DumpLayoutVerbosity {
    /// Includes more information in layout-dump, such as margin, border, and padding values for
    /// each box.
    Verbose,
    /// The most minimal layout-dump representation, including information such as box size, box
    /// type, xy position coordinates, and more.
    NonVerbose,
}

impl DumpLayoutVerbosity {
    pub fn to_cli_string(&self) -> String {
        // The CLI form of this flag is currently a boolean.
        match self {
            DumpLayoutVerbosity::Verbose => "1",
            DumpLayoutVerbosity::NonVerbose => "0",
        }
        .to_owned()
    }
}

pub fn dump_layout_tree_verbose(
    dump_layout_arg_matches: &ArgMatches,
) -> Option<DumpLayoutVerbosity> {
    try_get_bool(dump_layout_arg_matches, "verbose").map(|bool_verbose| match bool_verbose {
        true => DumpLayoutVerbosity::Verbose,
        false => DumpLayoutVerbosity::NonVerbose,
    })
}
