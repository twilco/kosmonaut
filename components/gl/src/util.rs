use crate::types::GLint;
use crate::{Gl, VERSION};
use std::ffi::{CStr, CString};

pub fn bool_from_glint(glint: GLint) -> bool {
    match glint {
        0 => false,
        1 => true,
        _ => panic!("could not convert GLint to bool: {}", glint),
    }
}

pub fn opengl_version(gl: &Gl) -> String {
    unsafe {
        let data = CStr::from_ptr(gl.GetString(VERSION) as *const _)
            .to_bytes()
            .to_vec();
        String::from_utf8(data).unwrap()
    }
}

pub(crate) fn create_whitespace_cstring(len: usize) -> CString {
    // `len` + 1 to give space for the nul-terminate byte at the end
    let mut buffer: Vec<u8> = Vec::with_capacity(len + 1);
    buffer.extend([b' '].iter().cycle().take(len));
    unsafe { CString::from_vec_unchecked(buffer) }
}
