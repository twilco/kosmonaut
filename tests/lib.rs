#![feature(custom_test_frameworks)]
#![test_runner(datatest::runner)]

use crate::util::CommandUnderTest;
use cli::DumpLayoutVerbosity;
use insta::assert_snapshot;
use std::ffi::OsStr;
use std::path::Path;

mod util;

mod layout;
mod style;

pub(crate) const LAYOUT_DUMP_INNER_WINDOW_WIDTH_PX: f32 = 1920.;
pub(crate) const LAYOUT_DUMP_INNER_WINDOW_HEIGHT_PX: f32 = 1080.;
/// Unless otherwise specified, run layout tests with a scale factor of 1.0.
pub(crate) const LAYOUT_DUMP_DEFAULT_SCALE_FACTOR: f32 = 1.0;
pub(crate) const LAYOUT_DUMP_DEFAULT_VERBOSITY: DumpLayoutVerbosity =
    DumpLayoutVerbosity::NonVerbose;

/// This automatically finds and runs any <filename>.reftest.html and <filename>.expected.html as
/// a reftest.  A reftest renders the given documents headlessly and performs a pixel-by-pixel
/// comparison of the results.
#[datatest::files("tests/websrc", {
reftest_html_file in r"^(.*).reftest.html",
expected_html_file = r"${1}.expected.html",
})]
fn auto_reftests(reftest_html_file: &Path, expected_html_file: &Path) {
    reftest_expect_similar(
        reftest_html_file.to_str().unwrap(),
        expected_html_file.to_str().unwrap(),
    )
}

/// Automatically discovers and runs dump-layout tests with the proper filename.
///
/// This will only ever run dump-layout with a set of default settings.  If your test requires any
/// sort of customization (e.g. changing scale factor, screen width / height, etc), you will need
/// to manually create a test to consume `dump_layout_cmd`.
#[datatest::files("tests/websrc", {
dump_layout_html_file in r"^(.*).dumplayout.html",
})]
fn auto_dump_layout_tests(dump_layout_html_file: &Path) {
    let mut dump_layout_cmd = dump_layout_cmd(DumpLayoutVerbosity::NonVerbose);
    dump_layout_cmd
        .arg(dump_layout_html_file.to_str().unwrap())
        .succeeds();
    snapshot_dump_layout_cmd(dump_layout_cmd, dump_layout_html_file);
}

pub fn snapshot_dump_layout_cmd(dump_layout_cmd: CommandUnderTest, corpus_html_file: &Path) {
    snapshot_dump_layout_cmd_verbose_scaled(
        dump_layout_cmd,
        corpus_html_file,
        LAYOUT_DUMP_DEFAULT_VERBOSITY,
        LAYOUT_DUMP_DEFAULT_SCALE_FACTOR,
    )
}

pub fn snapshot_dump_layout_cmd_verbose(
    dump_layout_cmd: CommandUnderTest,
    corpus_html_file: &Path,
    verbosity: DumpLayoutVerbosity,
) {
    snapshot_dump_layout_cmd_verbose_scaled(
        dump_layout_cmd,
        corpus_html_file,
        verbosity,
        LAYOUT_DUMP_DEFAULT_SCALE_FACTOR,
    )
}

pub fn snapshot_dump_layout_cmd_scaled(
    dump_layout_cmd: CommandUnderTest,
    corpus_html_file: &Path,
    scale_factor: f32,
) {
    snapshot_dump_layout_cmd_verbose_scaled(
        dump_layout_cmd,
        corpus_html_file,
        LAYOUT_DUMP_DEFAULT_VERBOSITY,
        scale_factor,
    )
}

pub fn snapshot_dump_layout_cmd_verbose_scaled(
    dump_layout_cmd: CommandUnderTest,
    corpus_html_file: &Path,
    verbosity: DumpLayoutVerbosity,
    scale_factor: f32,
) {
    let settings = dump_layout_insta_settings(corpus_html_file, verbosity, scale_factor);
    let (insta_settings, snapshot_filename) = (settings.insta_settings, settings.snapshot_filename);
    insta_settings.bind(|| {
        assert_snapshot!(snapshot_filename, dump_layout_cmd.stdout());
    });
}

/// Computes the directory to save the snapshot to based on the corpus.  The root test websrc
/// directory (which contains our corpus test files) is stripped off the front of the given
/// path, and the filename of this specific testcase is stripped off the end.
///
/// "tests/websrc/directional/rtl-horizontal-tb-block-boxes.layoutsnap.html"
///   becomes
/// "snapshots/directional"
pub fn compute_snapshot_dir(dump_layout_html_file: &Path) -> String {
    let file_name = dump_layout_html_file.file_name().unwrap();
    let path = dump_layout_html_file.to_str().unwrap();
    let trimmed_path = path
        .trim_start_matches("tests/websrc/")
        .trim_end_matches(file_name.to_str().unwrap())
        .trim_end_matches("/");
    format!("{}/{}", "snapshots", trimmed_path)
}

pub struct DumpLayoutInstaSettings {
    pub insta_settings: insta::Settings,
    pub snapshot_filename: String,
}

fn dump_layout_insta_settings(
    corpus_html_file_path: &Path,
    verbosity: DumpLayoutVerbosity,
    scale_factor: f32,
) -> DumpLayoutInstaSettings {
    let mut settings = insta::Settings::clone_current();
    settings.set_input_file(corpus_html_file_path);
    settings.set_snapshot_path(compute_snapshot_dir(corpus_html_file_path));
    settings.set_prepend_module_to_snapshot(false);
    let snapshot_suffix = match verbosity {
        LAYOUT_DUMP_DEFAULT_VERBOSITY if scale_factor == LAYOUT_DUMP_DEFAULT_SCALE_FACTOR => {
            "".to_owned()
        }
        verbosity if scale_factor == LAYOUT_DUMP_DEFAULT_SCALE_FACTOR => match verbosity {
            DumpLayoutVerbosity::Verbose => "verbose",
            LAYOUT_DUMP_DEFAULT_VERBOSITY => "",
        }
        .to_owned(),
        LAYOUT_DUMP_DEFAULT_VERBOSITY if scale_factor != LAYOUT_DUMP_DEFAULT_SCALE_FACTOR => {
            format!("scaled-{}", scale_factor)
        }
        _ if scale_factor != LAYOUT_DUMP_DEFAULT_SCALE_FACTOR => {
            format!("verbose-scaled-{}", scale_factor)
        }
        _ => panic!(
            "failure to create snapshot suffix with verbosity '{:?}' and scale_factor '{}'",
            verbosity, scale_factor
        ),
    };
    settings.set_snapshot_suffix(snapshot_suffix);
    let file_name = corpus_html_file_path.file_name().unwrap();
    DumpLayoutInstaSettings {
        insta_settings: settings,
        snapshot_filename: file_name.to_str().unwrap().to_owned(),
    }
}

pub fn reftest_expect_similar<S: AsRef<str> + AsRef<OsStr>>(file_path_one: S, file_path_two: S) {
    let (file_path_one, file_path_two): (&str, &str) =
        (file_path_one.as_ref(), file_path_two.as_ref());
    match run_similarity_cmd(file_path_one, file_path_two) {
        Ok(percent_similar) => {
            if percent_similar != 100.0f64 {
                panic!("reftest_expect_similar failure.  files '{}' and '{}' were only {}% similar (expected 100%)", file_path_one, file_path_two, percent_similar);
            }
        }
        Err(err) => panic!(
            "reftest_expect_similar *internal* failure (this is bad) with files '{}' and '{}'.  error message was: {}",
            file_path_one,
            file_path_two,
            err
        ),
    }
}

fn run_similarity_cmd<S: AsRef<str> + AsRef<OsStr>>(
    file_path_one: S,
    file_path_two: S,
) -> Result<f64, String> {
    let mut cmd = CommandUnderTest::new();
    cmd.arg("similarity");
    cmd.arg(file_path_one);
    cmd.arg(file_path_two);
    cmd.arg("--similarity-percent-only");
    cmd.succeeds();
    let stdout = cmd.stdout().trim();
    stdout.parse::<f64>().map_err(|_| stdout.trim().to_owned())
}

pub fn dump_layout_cmd_verbose_scaled(
    verbosity: DumpLayoutVerbosity,
    scale_factor: f32,
) -> CommandUnderTest {
    let mut cmd = CommandUnderTest::new();
    cmd.arg("dump-layout");
    cmd.arg("--width");
    cmd.arg(format!("{}", LAYOUT_DUMP_INNER_WINDOW_WIDTH_PX));
    cmd.arg("--height");
    cmd.arg(format!("{}", LAYOUT_DUMP_INNER_WINDOW_HEIGHT_PX));
    cmd.arg("--verbose");
    cmd.arg(format!("{}", verbosity.to_cli_string()));
    cmd.arg("--scale-factor");
    cmd.arg(format!("{}", scale_factor));
    cmd
}

pub fn dump_layout_cmd(verbosity: DumpLayoutVerbosity) -> CommandUnderTest {
    dump_layout_cmd_verbose_scaled(verbosity, LAYOUT_DUMP_DEFAULT_SCALE_FACTOR)
}
