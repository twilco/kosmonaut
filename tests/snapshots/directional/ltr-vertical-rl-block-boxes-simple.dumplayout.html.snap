---
source: tests/lib.rs
expression: dump_layout_cmd.stdout()
input_file: tests/websrc/directional/ltr-vertical-rl-block-boxes-simple.dumplayout.html

---
HTML BlockContainer at (822, 0) size 1098x1080
  BODY BlockContainer at (831, 9) size 1080x1062
    DIV BlockContainer at (1581, 39) size 300x200
    DIV BlockContainer at (1221, 39) size 300x200
    DIV BlockContainer at (861, 39) size 300x200

